gcc/d/decl.cc

```diff
@@ -1181,6 +1196,14 @@ get_symbol_decl (Declaration *decl)
 	    }
 	  else
 	    DECL_EXTERNAL (decl->csym) = 1;
+
+	  d_comdat_linkage (decl->csym);
+
+	  /* Normally the backend only emits COMDAT things when they are needed.
+	     If this decl is meant to be externally visible, then make sure that
+	     to mark it so that it is indeed needed.  */
+	  if (TREE_PUBLIC (decl->csym))
+	    mark_needed (decl->csym);
```

Comment:
This is where user declared symbols land.  Would be here where more fine grain control would happen.

Generally though, the front-end let us know more about the instantiation context of the decl, in particular:
1. Was this instantiated in CTFE?  Ideally these decls should be discarded by the front-end and never reach here.
2. Is this instantiated in a function or module context?  Or in C++ terminology, is this implicitly or explicitly instantiated?  As far as I can see, only explicit templates (`module foo; alias bar = Foo!42;`) really need all members emitted regardless of whether or not they are used.